router/http_test.go

```diff
@@ -339,6 +339,202 @@ func (s *S) TestHTTPHeadersFromClient(c *C) {
+			presentHeaders: []string{"Another-Option", "Custom-Conn-Header", "Keep-Alive"},
+		},
+		{
+			conn: "keep-alive",
+			// Keep-Alive header should be deleted because that's a conn-specific
+			// header here. Upgrade still gets deleted b/c Connection != "upgrade".
+			emptyHeaders:   []string{"Keep-Alive", "Transfer-Encoding", "Upgrade"},
+			presentHeaders: []string{"Another-Option", "Custom-Conn-Header"},
+		},
+		{
+			conn:           "custom-conn-header",
+			emptyHeaders:   []string{"Custom-Conn-Header", "Transfer-Encoding", "Upgrade"},
+			presentHeaders: []string{"Another-Option", "Keep-Alive"},
+		},
+		{ // test multiple connection-options
+			conn:           "custom-conn-header,   ,another-option   ",
+			emptyHeaders:   []string{"Another-Option", "Custom-Conn-Header", "Transfer-Encoding", "Upgrade"},
+			presentHeaders: []string{"Keep-Alive"},
+		},
+		{
+			// tcp/websocket path, all headers should be sent to backend (except
+			// Transfer-Encoding)
+			conn:              "upgrade",
+			upgradeFromClient: true,
+			emptyHeaders:      []string{"Transfer-Encoding"},
+			presentHeaders:    []string{"Custom-Conn-Header", "Keep-Alive", "Upgrade"},
+		},
+		{
+			// tcp/websocket path, all headers should be sent to backend (except
+			// Transfer-Encoding)
+			conn:              "upgrade, custom-conn-header,   ,another-option   ",
+			upgradeFromClient: true,
+			emptyHeaders:      []string{"Transfer-Encoding"},
+			presentHeaders:    []string{"Another-Option", "Custom-Conn-Header", "Keep-Alive", "Upgrade"},
+		},
+	}
+	for _, test := range tests {
+		testConnectionHeadersFromClient(c, l, srv, test.conn, test.emptyHeaders, test.presentHeaders)
+	}
+	for _, test := range tests {
+		testConnectionHeadersFromServer(c, l, srv, test.conn, test.upgradeFromClient, test.emptyHeaders, test.presentHeaders)
+	}
+}
+
+func testConnectionHeadersFromClient(
+	c *C,
+	l *httpListener,
+	srv *httptest.Server,
+	conn string,
+	emptyHeaders,
+	presentHeaders []string,
+) {
+	c.Logf("testing client with Connection: %q", conn)
+	srv.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		for _, k := range emptyHeaders {
+			c.Logf("checking if %q header is empty", k)
+			c.Assert(req.Header.Get(k), Equals, "")
+		}
+		for _, k := range presentHeaders {
+			c.Logf("checking if %q header is present", k)
+			c.Assert(req.Header.Get(k), Not(Equals), "")
+		}
+	})
+	req := newReq("http://"+l.Addr, "example.com")
+	if conn != "" {
+		req.Header.Set("Connection", conn)
+	}
+	req.Header.Set("Another-Option", "test-another-option")
+	req.Header.Set("Custom-Conn-Header", "test-custom-conn-header")
+	req.Header.Set("Keep-Alive", "test-keep-alive")
+	req.Header.Set("Transfer-Encoding", "test-transfer-encoding")
+	req.Header.Set("Upgrade", "test-upgrade")
+	res, err := httpClient.Do(req)
+	c.Assert(err, IsNil)
+	defer res.Body.Close()
+	c.Assert(res.StatusCode, Equals, 200)
+}
+
+func testConnectionHeadersFromServer(
```

Comment:
Any reason these are separate functions? Seems like they could just go inline in the for loop.